kb/how-to/add-test-generation.md

```diff
@@ -0,0 +1,478 @@
+
+```python
+#!/usr/bin/env python3
+import sys
+
+import scripts_path
+from mbedtls_dev import test_generation
+
+if __name__ == '__main__':
+    # Pass command line arguments and a description for the script
+    test_generation.main(sys.argv[1:], "Generate bignum tests.")
+```
+
+This script can now be run, although it will do nothing at this stage.
+Use the `--help` argument for usage details.
+
+```bash
+$ python tests/scripts/generate_bignum_ex_tests.py
+```
+
+### Generating a data file
+
+To generate a data file from the script, a subclass of `BaseTarget` must be defined,
+setting `target_basename` to the output file basename.
+This is a [file target class](#file-target-classes).
+Generally `target_basename` will be set to `"test_suite_xyz.generated"` for tests defined
+in `test_suite_xyz.function`.
+In the example, `metaclass=ABCMeta` is added to the class to explicitly indicate this is
+an abstract class.
+As Pylint does not recognize this as an abstract class, the `abstract-method` warning is
+disabled.
+
+For the example, add the following class:
+
+```python
+from abc import ABCMeta
+...
+class BignumTarget(test_generation.BaseTarget, metaclass=ABCMeta):
+    #pylint: disable=abstract-method
+    """Target for bignum test case generation."""
+    # Using .ex in this example to avoid clash with `generate_bignum_tests.py`
+    target_basename = "test_suite_mpi.ex.generated"
+```
+
+Running the script will now create `test_suite_mpi.ex.generated.data`, and running with
+`--list` will list the filename as a valid target.
+
+### Adding the script to the build system
+
+This is not required for the example, but is included here for completeness.
+When adding a new script to the build system, the changes required include:
+ - Adding the script to
+[/tests/scripts/check-generated-files.sh](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/check-generated-files.sh).
+ - Adding the script to
+[/scripts/make_generated_files.bat](https://github.com/Mbed-TLS/mbedtls/blob/development/scripts/make_generated_files.bat).
+ - Adding calls and targets in CMake and Make for the script and its generated files.
+The script will need to be called in the same places where `generate_bignum_tests.py` is
+currently called.
+
+## Generating test cases for a function
+
+To add test cases for a function, a concrete class derived from `BaseTarget` must be
+added to the script.
+This class must implement all attributes and methods of `BaseTarget` and will generate
+test cases.
+This is a [test function class](#test-function-classes).
+
+As an example, test case generation is added for the test function `mpi_add_mpi()`.
+The added class will be derived from `BignumTarget`, and can be added either to
+[tests/scripts/generate_bignum_tests.py](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/generate_bignum_tests.py)
+or the example script described in the previous section.
+
+### Creating the class
```

Comment:
For the sake of people who aren't reading the document linearly, please clarify “the class” in the title. Even just “Creating the function class” would be a lot clearer.